File: LeetcodeProblems/blind-75/graph/number-of-islands/solution.py
# Number of Islands - LeetCode #200

from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0
        
        m = len(grid)
        n = len(grid[0])
        count = 0
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    count += 1
        
        return count
    
    def dfs(self, grid: List[List[str]], i: int, j: int):
        m = len(grid)
        n = len(grid[0])
        
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != '1':
            return
        
        grid[i][j] = '0'  # Visited
        
        self.dfs(grid, i - 1, j)
        self.dfs(grid, i + 1, j)
        self.dfs(grid, i, j - 1)
        self.dfs(grid, i, j + 1)

File: LeetcodeProblems/blind-75/graph/number-of-islands/solution.js
// Number of Islands - LeetCode #200

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    if (!grid || grid.length === 0) return 0;
    
    const m = grid.length;
    const n = grid[0].length;
    let count = 0;
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === '1') {
                dfs(grid, i, j);
                count++;
            }
        }
    }
    
    return count;
};

function dfs(grid, i, j) {
    const m = grid.length;
    const n = grid[0].length;
    
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] !== '1') return;
    
    grid[i][j] = '0'; // Visited
    
    dfs(grid, i - 1, j);
    dfs(grid, i + 1, j);
    dfs(grid, i, j - 1);
    dfs(grid, i, j + 1);
}
